From 8e419be7dec5f495de7adf116a941660fda7f4d6 Mon Sep 17 00:00:00 2001
From: Alexander Shishkin <alex@ash.works>
Date: Fri, 26 Sep 2025 17:41:02 +0300
Subject: [PATCH 1/3] std_vorbis: Remove a redundant pointer comparison

The check for `ptr + unsigned int < ptr` is always false, and the
compilers are starting to catch up. Remove it.

Also, update replace emscripten's deprecated version macros with the
current ones.

Signed-off-by: Alexander Shishkin <alex@ash.works>
---
 extras/stb_vorbis.c | 2 +-
 miniaudio.h         | 4 ++--
 2 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/extras/stb_vorbis.c b/extras/stb_vorbis.c
index 3e5c2504..07e2a17a 100644
--- a/extras/stb_vorbis.c
+++ b/extras/stb_vorbis.c
@@ -1401,7 +1401,7 @@ static int set_file_offset(stb_vorbis *f, unsigned int loc)
    #endif
    f->eof = 0;
    if (USE_MEMORY(f)) {
-      if (f->stream_start + loc >= f->stream_end || f->stream_start + loc < f->stream_start) {
+      if (f->stream_start + loc >= f->stream_end) {
          f->stream = f->stream_end;
          f->eof = 1;
          return 0;
diff --git a/miniaudio.h b/miniaudio.h
index 24e676bb..441f8de6 100644
--- a/miniaudio.h
+++ b/miniaudio.h
@@ -41495,11 +41495,11 @@ Web Audio Backend
 #ifdef MA_HAS_WEBAUDIO
 #include <emscripten/emscripten.h>
 
-#if (__EMSCRIPTEN_major__ > 3) || (__EMSCRIPTEN_major__ == 3 && (__EMSCRIPTEN_minor__ > 1 || (__EMSCRIPTEN_minor__ == 1 && __EMSCRIPTEN_tiny__ >= 32)))
+#if (__EMSCRIPTEN_MAJOR__ > 3) || (__EMSCRIPTEN_MAJOR__ == 3 && (__EMSCRIPTEN_MINOR__ > 1 || (__EMSCRIPTEN_MINOR__ == 1 && __EMSCRIPTEN_TINY__ >= 32)))
     #include <emscripten/webaudio.h>
     #define MA_SUPPORT_AUDIO_WORKLETS
 
-    #if (__EMSCRIPTEN_major__ > 3) || (__EMSCRIPTEN_major__ == 3 && (__EMSCRIPTEN_minor__ > 1 || (__EMSCRIPTEN_minor__ == 1 && __EMSCRIPTEN_tiny__ >= 70)))
+    #if (__EMSCRIPTEN_MAJOR__ > 3) || (__EMSCRIPTEN_MAJOR__ == 3 && (__EMSCRIPTEN_MINOR__ > 1 || (__EMSCRIPTEN_MINOR__ == 1 && __EMSCRIPTEN_TINY__ >= 70)))
         #define MA_SUPPORT_AUDIO_WORKLETS_VARIABLE_BUFFER_SIZE
     #endif
 #endif
-- 
2.54.0 (Apple Git-157)


From f04464add5cb7132bf4405166b6a224e021f6ee4 Mon Sep 17 00:00:00 2001
From: prevter <prevter@gmail.com>
Date: Fri, 18 Sep 2026 14:11:50 +0300
Subject: [PATCH 2/3] Zero-initialize EmscriptenWebAudioCreateAttributes

(cherry picked from commit 78c8a43a5f99980fa4cbf34ac5720c3f77073c4d)
Signed-off-by: Alexander Shishkin <alex@ash.works>
---
 miniaudio.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/miniaudio.h b/miniaudio.h
index 441f8de6..88596481 100644
--- a/miniaudio.h
+++ b/miniaudio.h
@@ -41996,6 +41996,7 @@ static ma_result ma_device_init__webaudio(ma_device* pDevice, const ma_device_co
         ma_audio_worklet_thread_initialized_data* pInitParameters;
         void* pStackBuffer;
 
+        MA_ZERO_OBJECT(&audioContextAttributes);
         if (pConfig->performanceProfile == ma_performance_profile_conservative) {
             audioContextAttributes.latencyHint = MA_WEBAUDIO_LATENCY_HINT_PLAYBACK;
         } else {
-- 
2.54.0 (Apple Git-157)


From b7a6f640c0f8ad9bf938051169c21c06afe83ada Mon Sep 17 00:00:00 2001
From: Alexander Shishkin <alex@ash.works>
Date: Tue, 20 Oct 2026 03:26:43 +0300
Subject: [PATCH 3/3] Web Audio: Silence the worklet until the device is
 initialized

ma_device_init__webaudio() connects the AudioWorkletNode to the
destination from the worklet's creation callback, while ma_device_init()
waits in emscripten_sleep(); the internal period size is only set after
that, by ma_device_post_init(). A process callback that runs in between
sees a device that isn't started and silences internalPeriodSizeInFrames
frames of it, which is 0, and returns EM_TRUE. Emscripten then sends
out its output views as they are: the top of the worklet's stack, which
is ma_aligned_malloc()ed and never cleared. Where the browser runs the
callback in that window, that's a burst of full scale garbage as soon as
the audio starts.

Size the silence by the frames Emscripten passes in.

Signed-off-by: Alexander Shishkin <alex@ash.works>
---
 miniaudio.h | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/miniaudio.h b/miniaudio.h
index 88596481..db7e444b 100644
--- a/miniaudio.h
+++ b/miniaudio.h
@@ -41765,11 +41765,17 @@ static EM_BOOL ma_audio_worklet_process_callback__webaudio(int inputCount, const
 
     /*
     If this is called by the device has not yet been started we need to return early, making sure we output silence to
-    the output buffer.
+    the output buffer. The silence is sized by the frame itself: the node is connected to the destination before
+    ma_device_init() sets the internal period size, so the callback can run with a frameCount of 0 and the outputs,
+    which Emscripten keeps on the worklet's stack, would go out as they are.
     */
     if (ma_device_get_state(pDevice) != ma_device_state_started) {
         for (int i = 0; i < outputCount; i += 1) {
-            MA_ZERO_MEMORY(pOutputs[i].data, pOutputs[i].numberOfChannels * frameCount * sizeof(float));
+            #if defined(MA_SUPPORT_AUDIO_WORKLETS_VARIABLE_BUFFER_SIZE)
+            MA_ZERO_MEMORY(pOutputs[i].data, pOutputs[i].numberOfChannels * pOutputs[i].samplesPerChannel * sizeof(float));
+            #else
+            MA_ZERO_MEMORY(pOutputs[i].data, pOutputs[i].numberOfChannels * 128 * sizeof(float));
+            #endif
         }
 
         return EM_TRUE;
-- 
2.54.0 (Apple Git-157)

